[ROCm][CK][Inductor] Enable addmm for CK backend to gemm max autotune

tenpercent opened the pull request:
Add functional support for torch.addmm with CK backend. See also #125453 

# Implementation details
1. It turns out we can use the same template between addmm and matmul; essentially, matmul is addmm with empty bias
2. The Python generator in CK was updated to generate the shared cpp template. The pip package can be installed from `pip install git+https://github.com/rocm/composable_kernel@add-addmm` and will be merged into `develop` branch after this PR lands to avoid breaking the current matmul

# Testing
`pytest test/inductor/test_ck_backend.py -k addmm`

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @zjing14 @asroy @nmacchioni @aska-0096

tenpercent committed:
try adding addmm

tenpercent committed:
merge addmm with matmul

tenpercent committed:
rename supported_arch to ck_supported_arch in rocm config

aakhundov commented on torch/_inductor/kernel/mm.py:
```diff
@@ -339,6 +339,15 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
                 beta=beta,
             )
 
+    if use_ck_template(layout, m, n, k):
```
So I've quickly scrolled through the `use_ck_template` and didn't notice a check that all input shapes are static. Should we add the check (and maybe others if relevant) here, similar to the CUTLASS backend above?

tenpercent replied:
I believe the check on `V.graph.sizevars.size_hint(m * n * k, fallback=-1)` falls back when the shape is dynamic?

zoranzhao replied:
Yeah, I think it should be fine: https://github.com/pytorch/pytorch/blob/c4bf4005d1eafc755f0cdd2379ade0bc3bb3902c/torch/_inductor/sizevars.py#L484

aakhundov commented on torch/_inductor/codegen/rocm/rocm_kernel.py:
```diff
@@ -211,6 +214,21 @@ def size(
         val = sympy_product(sizes)
         return cexpr(self.rename_indexing(val))
 
+    def contiguous_stride(self, node: IRNode, default_value: int = 0) -> str:
```
Nit: maybe `get_contiguous_stride`?

tenpercent replied:
but: size() and stride() in the CUTLASS backend. I think 'get' is sort of implied here and there

aakhundov commented on torch/_inductor/codegen/rocm/rocm_kernel.py:
```diff
@@ -211,6 +214,21 @@ def size(
+        if node is None:
+            return str(default_value)
+
+        contiguous_stride = functools.reduce(lambda a, b: max(b, a), node.get_stride(), sympy.Integer(0))
```
Does Python `max` work well with symbolic shapes? Or should we use something internal (from `experimental/symbolic_shapes.py`)?

tenpercent replied:
these lines are the inlined version of symbolic product for the tensor size from its dimensions. Not sure about how the symbolic shape looks like to test it

aakhundov commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -231,9 +255,14 @@ def emit_ck_instance(self, op: "CKGemmOperation"):
-                )
+                if "ds" in field_name:  # element type and layout for bias
+                    template_params.append(
+                        f"/* {field_name} */ Tuple<{', '.join(map(str, iter(field_value)))}>"
+                    )
+                else:  # tile shape
+                    template_params.append(
+                        f"/* {field_name} */ S<{', '.join(map(str, iter(field_value)))}>"
```
Hoist the `', '.join(map(str, iter(field_value)))`? And maybe `template_params.append`.

aakhundov commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -251,10 +280,20 @@ def render(self, kernel: ROCmTemplateKernel, op: "CKGemmOperation", **kwargs) ->
+        if Bias:
+            op.ds_layouts = (torch_layout_to_ck_layout(Bias.get_layout()),)
+            op.ds_element_dtypes = ((self._TORCH_DTYPE_TO_CK[Bias.get_layout().dtype]),)
+            op.c_elementwise_op = "ScaledAdd"
```
Genericness of this implies that we can, potentially, support more elaborate epilogue fusion?

tenpercent replied:
Yes, the template can support at least elementwise fusion, need to think about more complex patterns

aakhundov commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -251,10 +280,20 @@ def render(self, kernel: ROCmTemplateKernel, op: "CKGemmOperation", **kwargs) ->
+
+        op = copy.deepcopy(op)
+
+        if Bias:
```
Maybe check `if Bias is not None` to be on the safe side? Also everywhere below.

aakhundov commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -282,16 +321,19 @@ def render(self, kernel: ROCmTemplateKernel, op: "CKGemmOperation", **kwargs) ->
+            bias_element_dtype=op.ds_element_dtypes[0] if Bias else '',
+            a_stride=kernel.contiguous_stride(X),
+            b_stride=kernel.contiguous_stride(W),
+            c_stride=kernel.contiguous_stride(Y),
+            d_stride=kernel.contiguous_stride(Bias),
```
So in CK, C is output and D is bias? I think, in CUTLASS it's the other way around.

tenpercent replied:
Yes. Well there is inconsistency sometimes that C is intermediate result of the gemm proper, Ds are 'bias'es and E is the output

aakhundov commented:
@tenpercent the linter seems failing, please take a look.

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -137,9 +158,12 @@ def filter_op(self, op: "CKGemmOperation"):
             T.get_layout() for T in [*self.input_nodes, self.output_node]
-        )
+        ]
+        X_meta = metas[0]
+        W_meta = metas[1]
+        Y_meta = metas[-1]
```
Yes, len(inputs) can be 2 or 3 whether the op is matmul or addmm, respectively

tenpercent commented on test/inductor/test_ck_backend.py:
```diff
@@ -165,6 +165,41 @@ def mm(a, b):
+            Y_compiled = addmm(x, a, b, alpha, beta)
+            Y_eager = torch.addmm(x, a, b, alpha=alpha, beta=beta)
+
+            torch.testing.assert_close(Y_compiled, Y_eager)
```
It seems tricky to implement. The place where the generated files are stored seems to be runtime specific. Also haven't seen the CUTLASS backend tests doing it

tenpercent commented on test/inductor/test_ck_backend.py:
```diff
@@ -165,6 +165,41 @@ def mm(a, b):
+        with config.patch(
+            {
+                "max_autotune": True,
+                "autotune_in_subproc": True,
```
At the moment it is just chosen to mimic a corresponding CUTLASS backend test. Probably we should add the another test as well

zoranzhao commented on torch/_inductor/kernel/mm.py:
```diff
@@ -339,6 +340,15 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
                 beta=beta,
             )
 
+    if use_ck_template(layout, m, n, k):
```
Shall we do 
`if static_shape and is_nonzero and use_ck_template(layout, m, n, k):` ?

zoranzhao commented on torch/_inductor/codegen/rocm/ck_addmm_template.py:
```diff
@@ -0,0 +1,503 @@
+        return None
+
+
+class CKAddMMTemplate(CKTemplate):
```
nit, I am thinking whether we can add some unit test for this template class

tenpercent replied:
This class was removed, now CKGemmTemplate supports both matmul and addmm with a single template on CK side

zoranzhao reviewed (commented):
Left some comments

tenpercent commented:
@pytorchbot rebase -s

tenpercent committed:
run lintrunner

tenpercent committed:
disable ck templplate for dynamic shapes

tenpercent committed:
maybe fix type lint

tenpercent committed:
cleanup tuple/sequence rendering

tenpercent committed:
add UT for autotuning matmul in same process

tenpercent committed:
refactor bias is not none check

tenpercent committed:
remove unused imports

tenpercent committed:
manual lints

tenpercent committed:
newlines

tenpercent committed:
manual lints

tenpercent committed:
add cdna2 to arch filter

tenpercent commented:
I fixed the linter and addressed the comments by this point; however, the test started getting numerical mismatches for addmm (ok for matmul) so I need to investigate it

tenpercent committed:
fix 1d bias case

tenpercent committed:
debugging~

tenpercent committed:
tweak precision reduction in test case for consistency with other UTs

tenpercent committed:
rename index types to mirror the ck example for gemm with bias

tenpercent commented:
It seems the failing test case is turned off for the cutlass backend (as stride[-1] == 0 for broadcasted last dim)

tenpercent committed:
fix lints

tenpercent committed:
fix lints

tenpercent committed:
add comment

tenpercent committed:
try to leave better comments about strides for bias. also max sympyfication

tenpercent committed:
replace bespoke scaledadd with stock bilinear epilogue

tenpercent committed:
convert set to list, avoiding conflict with #131921

tenpercent committed:
Merge remote-tracking branch 'upstream/viable/strict' into ck-addmm

tenpercent committed:
Merge branch 'main' into ck-addmm

tenpercent committed:
disable (M,1) broadcasted bias

tenpercent committed:
make unit test pass; TBD to debug numerics

tenpercent committed:
Merge branch 'ck-addmm' of https://github.com/tenpercent/pytorch into ck-addmm

tenpercent committed:
add a comment about numerics

tenpercent committed:
fix bias accumulation dtype

tenpercent committed:
lint

tenpercent committed:
fix bias broadcasted along the last dim

tenpercent committed:
lint

tenpercent committed:
tweak k in addmm unit test

tenpercent committed:
update comment

chenyang78 commented on torch/_inductor/codegen/rocm/rocm_kernel.py:
```diff
@@ -210,6 +213,34 @@ def size(
 
+    def contiguous_stride(self, node: Optional[IRNode], default_value: int = 0) -> str:
+        """
+        Hook called from template call to get the contiguous stride of an arg.
```
Could you elaborate on what ```contiguous stride``` means? For example, does it refer to the stride of a particular dimension? It would be great if we could have some example to explain what it stands for. Thanks.

tenpercent replied:
Well the motivation here is, having a torch tensor metadata, provide the stride parameter which CK template expects when constructing the gemm argument. Which is the size of the contiguous dimension in all cases except for the broadcasted bias (either column and row wise) where it is 0.

chenyang78 commented on torch/_inductor/config.py:
```diff
@@ -949,10 +949,9 @@ class rocm:
     # Processor name reference: https://llvm.org/docs/AMDGPUUsage.html#processors
-    # Keep it ordered, unordered set can cause spurious inductor cache misses
-    supported_arch: List[str] = ["gfx940", "gfx941", "gfx942"]
+    ck_supported_arch: List[str] = ["gfx90a"] + ["gfx940", "gfx941", "gfx942"]
```
super nit - perhaps just ```["gfx90a", "gfx940", "gfx941", "gfx942"]```?

tenpercent replied:
Yep. Here I was trying to make it easier to understand which triplet corresponds to which arch without an extra trip to the triplet name reference. I think I can make it more explicit

tenpercent commented on torch/_inductor/utils.py:
```diff
@@ -1126,7 +1126,8 @@ def use_ck_template(layout, m, n, k):
     requested_supported_archs = [
-        requested_archs[k] for k in requested_archs.keys() & config.rocm.supported_arch
+        requested_archs[k]
+        for k in requested_archs.keys() & config.rocm.ck_supported_arch
```
I believe lintrunner won't let me get away with the old formatting. Can try reverting and re-running the CI though

tenpercent replied:
What changed is `supported_arch` -> `ck_supported_arch` which got changed in order to avoid leaving impression that rocm pytorch works only on MI300; this is the ck inductor backend specific flag

chenyang78 replied:
I see. Thanks!

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -44,31 +45,32 @@ class CKGemmTemplate(CKTemplate):
-        constexpr auto StrideA = std::is_same_v<{{a_layout}}, Row> ? K : M;
-        constexpr auto StrideB = std::is_same_v<{{b_layout}}, Row> ? N : K;
-        constexpr auto StrideC = std::is_same_v<{{c_layout}}, Row> ? N : M;
-        constexpr auto KBatch = 1; // split k into batches
```
It's supported in the old CK template (without multiple D) but not supported yet in the CK template with multiple D. Once it's supported, it's going to be yet another hyperparameter to keep in mind when selecting the best template - was hardcoded as 1 initially as it wasn't accounted for in the search

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -44,31 +45,32 @@ class CKGemmTemplate(CKTemplate):
-            {{b_elementwise_op}} {},
-            {{c_elementwise_op}} {}
+            std::array<ck::index_t, {{1 if has_bias else 0}}>{ {{'StrideD' if has_bias else ''}} },
+            StrideC, // == StrideE
```
I think in CK terminology stride is for what leading dimension is in BLAS. So it's output's leading dimension. But naming is a bit inconsistent at this point so some variable names referring to C (the matmul result) are in fact about E (the result after adding bias). Bias is D.

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -252,10 +258,35 @@ def render(self, kernel: ROCmTemplateKernel, op: "CKGemmOperation", **kwargs) ->
+            op.c_elementwise_op = "Bilinear"
+            # this dtype is also used for adding bias to matmul result
+            # before converting down to the result dtype
+            op.c_shuffle_dtype = "F32"
```
Hmm, good point, logically it should be set to the accumulation dtype

tenpercent commented on torch/_inductor/codegen/rocm/ck_universal_gemm_template.py:
```diff
@@ -252,10 +258,35 @@ def render(self, kernel: ROCmTemplateKernel, op: "CKGemmOperation", **kwargs) ->
+
+        op = copy.deepcopy(op)
+
+        op.c_shuffle_block_transfer_scalar_per_vector_n_per_block = (
```
The op object is the result of parsing a CK template without multiple D, and then is used to render a template with multiple D. Most of the parameters are the same, but this one needs to be adjusted. First element of the tuple relates to gemm result, and second element of the tuple relates to bias (D). In case of several biases fused together, it will contain one element for C as well as each of the Ds.

tenpercent committed:
add comments

tenpercent committed:
rename and simplify ck strides to leading dims

tenpercent committed:
update config

chenyang78 reviewed (commented):
Please fix the lint issues. Also, seems there are some rocm ci failures. Thanks!

tenpercent committed:
lint

tenpercent commented:
@pytorchbot rebase -s

tenpercent commented:
@pytorchbot label ciflow/trunk

tenpercent committed:
Merge remote-tracking branch 'upstream/viable/strict' into ck-addmm

tenpercent commented:
@pytorchbot merge
